onnxruntime/core/framework/session_state.h

```diff
@@ -59,6 +60,93 @@ struct MemoryPatternGroup;
 class MemoryInfo;
 #endif
 
+struct PartialGraphExecutionManager {
+  PartialGraphExecutionManager() {
+    graph_runs_counter_ = 0;
+  }
+
+  int64_t GetNewPartialGraphRunId() {
+    std::lock_guard<OrtMutex> lock(lock_);
+    int64_t new_run_id;
+    if (available_run_ids_.size() > 0) {
+      new_run_id = available_run_ids_.back();
+      available_run_ids_.pop_back();
+    } else {
+      new_run_id = graph_runs_counter_++;
+    }
+
+    ORT_ENFORCE(to_be_executed_runs_.find(new_run_id) == to_be_executed_runs_.end());
+
+    to_be_executed_runs_.insert(new_run_id);
+    return new_run_id;
+  }
+
+  bool IsValidRunId(int64_t run_id) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it_in_progress_runs = graph_runs_.find(run_id);
+    auto it_to_be_executed = to_be_executed_runs_.find(run_id);
+
+    return ((it_in_progress_runs != graph_runs_.end()) ^ (it_to_be_executed != to_be_executed_runs_.end()));
+  }
+
+  bool IsToBeExecutedRunId(int64_t run_id) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = to_be_executed_runs_.find(run_id);
+    return it != to_be_executed_runs_.end();
+  }
+
+  void CancelPartialGraphRun(int64_t run_id) {
+    //REVIEW(codemzs): Does it make sense to cancel a partial run that was not "prepared". May be?
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = graph_runs_.find(run_id);
+
+    ORT_ENFORCE(it != graph_runs_.end());
+
+    available_run_ids_.push_back(run_id);
+    graph_runs_.erase(it);
+  }
+
+  void PreparePartialRun(std::shared_ptr<ExecutionFrame> frame, int64_t run_id, size_t program_counter) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = graph_runs_.find(run_id);
+
+    ORT_ENFORCE(it == graph_runs_.end());
+
+    auto it_to_be_executed = to_be_executed_runs_.find(run_id);
+
+    ORT_ENFORCE(it_to_be_executed != to_be_executed_runs_.end());
+
+    to_be_executed_runs_.erase(it_to_be_executed);
+
+    graph_runs_.insert(std::make_pair(run_id, std::make_pair(frame, program_counter)));
+  }
+
+  std::pair<std::shared_ptr<onnxruntime::ExecutionFrame>, size_t> GetPartialRun(int64_t run_id) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = graph_runs_.find(run_id);
+
+    ORT_ENFORCE(it != graph_runs_.end());
+
+    return it->second;
+  }
+
+  void UpdatePartialRunProgramCounter(int64_t run_id, size_t program_counter) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = graph_runs_.find(run_id);
+
+    ORT_ENFORCE(it != graph_runs_.end());
+
+    it->second.second = program_counter;
+  }
+
+ private:
+  OrtMutex lock_;
+  std::map<int64_t, std::pair<std::shared_ptr<onnxruntime::ExecutionFrame>, size_t>> graph_runs_;
+  std::vector<int64_t> available_run_ids_;
+  int64_t graph_runs_counter_;
+  std::unordered_set<int64_t> to_be_executed_runs_;
```

Comment:
>to_be_executed_runs_ [](start = 30, length = 20)

I am wondering would it be better if you merge graph_runs_ and  to_be_executed_runs_ into one.

you can use unassigned optional value to represent to_be_executed_runs_. 
see optional.h

Reply:
and you would need to only check on place for any lookup/isvalid check. 

---
In reply to: [602599956](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r602599956) [](ancestors = 602599956)